File: sdks/go/transport/codec.go
package transport

import (
	"bytes"
	"encoding/binary"
	"fmt"

	"github.com/vmihailenco/msgpack/v5"
)

// Wire codec for the Nexus RPC protocol.
//
// The server-side types live in `nexus_protocol::rpc::types`.
// rmp-serde encodes Rust enums using the externally-tagged
// representation by default, which means:
//
//   - Unit variants (`NexusValue::Null`) → the string "Null".
//   - Data-bearing variants (`NexusValue::Str("hi")`) → a single-key
//     map {"Str": "hi"}.
//   - `Result<T, E>` → {"Ok": v} / {"Err": s}.
//
// ToWire / FromWire translate between the tagged Go type and the
// on-wire shape. EncodeRequestFrame / DecodeResponseBody handle the
// `u32 LE length prefix + MessagePack` frame format.

// ToWire encodes a [NexusValue] into its on-wire (pre-MessagePack)
// Go shape — either the literal string "Null" or a single-key map.
func ToWire(v NexusValue) any {
	switch v.Kind {
	case KindNull:
		return "Null"
	case KindBool:
		return map[string]any{"Bool": v.Value}
	case KindInt:
		return map[string]any{"Int": v.Value}
	case KindFloat:
		return map[string]any{"Float": v.Value}
	case KindBytes:
		return map[string]any{"Bytes": v.Value}
	case KindStr:
		return map[string]any{"Str": v.Value}
	case KindArray:
		arr := v.Value.([]NexusValue)
		out := make([]any, len(arr))
		for i, e := range arr {
			out[i] = ToWire(e)
		}
		return map[string]any{"Array": out}
	case KindMap:
		pairs := v.Value.([]MapEntry)
		out := make([]any, len(pairs))
		for i, p := range pairs {
			out[i] = []any{ToWire(p.Key), ToWire(p.Value)}
		}
		return map[string]any{"Map": out}
	}
	panic(fmt.Sprintf("unknown NexusValue kind '%s'", v.Kind))
}

// FromWire decodes the wire-level Go shape back into a tagged [NexusValue].
//
// Accepts both the tagged shape ({"Str": "hi"}) and un-tagged primitives
// (bare bool / int / string / []byte / []any) so the codec tolerates
// servers (or tests) that emit the raw msgpack form directly.
func FromWire(raw any) (NexusValue, error) {
	switch x := raw.(type) {
	case nil:
		return NxNull(), nil
	case string:
		if x == "Null" {
			return NxNull(), nil
		}
		return NxStr(x), nil
	case bool:
		return NxBool(x), nil
	case int:
		return NxInt(int64(x)), nil
	case int64:
		return NxInt(x), nil
	case int32:
		return NxInt(int64(x)), nil
	case int16:
		return NxInt(int64(x)), nil
	case int8:
		return NxInt(int64(x)), nil
	case uint:
		return NxInt(int64(x)), nil
	case uint64:
		return NxInt(int64(x)), nil
	case uint32:
		return NxInt(int64(x)), nil
	case uint16:
		return NxInt(int64(x)), nil
	case uint8:
		return NxInt(int64(x)), nil
	case float32:
		return NxFloat(float64(x)), nil
	case float64:
		return NxFloat(x), nil
	case []byte:
		return NxBytes(x), nil
	case []any:
		out := make([]NexusValue, len(x))
		for i, e := range x {
			v, err := FromWire(e)
			if err != nil {
				return NexusValue{}, err
			}
			out[i] = v
		}
		return NxArray(out), nil
	case map[string]any:
		if len(x) != 1 {
			return NexusValue{}, fmt.Errorf(
				"decode: expected single-key tagged NexusValue, got %d keys", len(x),
			)
		}
		for tag, payload := range x {
			return fromWireTagged(tag, payload)
		}
	}
	return NexusValue{}, fmt.Errorf("decode: unexpected NexusValue wire type %T", raw)
}

func fromWireTagged(tag string, payload any) (NexusValue, error) {
	switch tag {
	case "Null":
		return NxNull(), nil
	case "Bool":
		b, ok := payload.(bool)
		if !ok {
			return NexusValue{}, fmt.Errorf("decode: Bool payload must be bool")
		}
		return NxBool(b), nil
	case "Int":
		switch v := payload.(type) {
		case int:
			return NxInt(int64(v)), nil
		case int8:
			return NxInt(int64(v)), nil
		case int16:
			return NxInt(int64(v)), nil
		case int32:
			return NxInt(int64(v)), nil
		case int64:
			return NxInt(v), nil
		case uint:
			return NxInt(int64(v)), nil
		case uint8:
			return NxInt(int64(v)), nil
		case uint16:
			return NxInt(int64(v)), nil
		case uint32:
			return NxInt(int64(v)), nil
		case uint64:
			return NxInt(int64(v)), nil
		}
		return NexusValue{}, fmt.Errorf("decode: Int payload must be integer, got %T", payload)
	case "Float":
		switch v := payload.(type) {
		case float32:
			return NxFloat(float64(v)), nil
		case float64:
			return NxFloat(v), nil
		case int64:
			return NxFloat(float64(v)), nil
		case int:
			return NxFloat(float64(v)), nil
		}
		return NexusValue{}, fmt.Errorf("decode: Float payload must be numeric, got %T", payload)
	case "Bytes":
		switch v := payload.(type) {
		case []byte:
			return NxBytes(v), nil
		case string:
			return NxBytes([]byte(v)), nil
		case []any:
			out := make([]byte, len(v))
			for i, e := range v {
				n, ok := e.(uint8)
				if !ok {
					return NexusValue{}, fmt.Errorf("decode: Bytes element must be uint8")
				}
				out[i] = n
			}
			return NxBytes(out), nil
		}
		return NexusValue{}, fmt.Errorf("decode: Bytes payload must be bytes, got %T", payload)
	case "Str":
		s, ok := payload.(string)
		if !ok {
			return NexusValue{}, fmt.Errorf("decode: Str payload must be string")
		}
		return NxStr(s), nil
	case "Array":
		arr, ok := payload.([]any)
		if !ok {
			return NexusValue{}, fmt.Errorf("decode: Array payload must be list")
		}
		out := make([]NexusValue, len(arr))
		for i, e := range arr {
			v, err := FromWire(e)
			if err != nil {
				return NexusValue{}, err
			}
			out[i] = v
		}
		return NxArray(out), nil
	case "Map":
		pairsRaw, ok := payload.([]any)
		if !ok {
			return NexusValue{}, fmt.Errorf("decode: Map payload must be list")
		}
		pairs := make([]MapEntry, len(pairsRaw))
		for i, p := range pairsRaw {
			kv, ok := p.([]any)
			if !ok || len(kv) != 2 {
				return NexusValue{}, fmt.Errorf("decode: Map entry must be [key, value] pair")
			}
			k, err := FromWire(kv[0])
			if err != nil {
				return NexusValue{}, err
			}
			v, err := FromWire(kv[1])
			if err != nil {
				return NexusValue{}, err
			}
			pairs[i] = MapEntry{Key: k, Value: v}
		}
		return NxMap(pairs), nil
	}
	return NexusValue{}, fmt.Errorf("decode: unknown NexusValue tag '%s'", tag)
}

// RpcRequest is the wire-level request frame.
type RpcRequest struct {
	ID      uint32
	Command string
	Args    []NexusValue
}

// RpcResponse is the decoded wire-level response.
type RpcResponse struct {
	ID  uint32
	OK  bool
	Val NexusValue // valid when OK=true
	Err string     // valid when OK=false
}

// Unwrap returns the response value or an error.
func (r RpcResponse) Unwrap() (NexusValue, error) {
	if !r.OK {
		return NexusValue{}, fmt.Errorf("server: %s", r.Err)
	}
	return r.Val, nil
}

// EncodeRequestFrame encodes a request into a length-prefixed
// MessagePack frame: `u32_le(body_len) ++ msgpack(body)`.
func EncodeRequestFrame(req RpcRequest) ([]byte, error) {
	args := make([]any, len(req.Args))
	for i, a := range req.Args {
		args[i] = ToWire(a)
	}
	body, err := msgpack.Marshal(map[string]any{
		"id":      req.ID,
		"command": req.Command,
		"args":    args,
	})
	if err != nil {
		return nil, err
	}
	out := make([]byte, 4+len(body))
	binary.LittleEndian.PutUint32(out[:4], uint32(len(body)))
	copy(out[4:], body)
	return out, nil
}

// DecodeResponseBody decodes a response body — the MessagePack bytes
// AFTER the length prefix. The caller is responsible for reading
// exactly `length` bytes off the wire before handing them here.
func DecodeResponseBody(body []byte) (RpcResponse, error) {
	dec := msgpack.NewDecoder(bytes.NewReader(body))
	dec.UseLooseInterfaceDecoding(true)
	var raw map[string]any
	if err := dec.Decode(&raw); err != nil {
		return RpcResponse{}, fmt.Errorf("decode: response must be a map: %w", err)
	}
	idRaw, ok := raw["id"]
	if !ok {
		return RpcResponse{}, fmt.Errorf("decode: response missing 'id'")
	}
	id, err := asUint32(idRaw)
	if err != nil {
		return RpcResponse{}, err
	}
	resultRaw, ok := raw["result"]
	if !ok {
		return RpcResponse{}, fmt.Errorf("decode: response missing 'result'")
	}
	resultMap, ok := resultRaw.(map[string]any)
	if !ok || len(resultMap) != 1 {
		return RpcResponse{}, fmt.Errorf("decode: Result must be a single-key tagged map")
	}
	for tag, payload := range resultMap {
		switch tag {
		case "Ok":
			val, err := FromWire(payload)
			if err != nil {
				return RpcResponse{}, err
			}
			return RpcResponse{ID: id, OK: true, Val: val}, nil
		case "Err":
			msg, _ := payload.(string)
			return RpcResponse{ID: id, OK: false, Err: msg}, nil
		default:
			return RpcResponse{}, fmt.Errorf("decode: Result must be 'Ok' or 'Err', got '%s'", tag)
		}
	}
	return RpcResponse{}, fmt.Errorf("decode: empty result map")
}

func asUint32(v any) (uint32, error) {
	switch n := v.(type) {
	case uint32:
		return n, nil
	case uint64:
		return uint32(n), nil
	case uint:
		return uint32(n), nil
	case int64:
		return uint32(n), nil
	case int:
		return uint32(n), nil
	case int32:
		return uint32(n), nil
	}
	return 0, fmt.Errorf("decode: 'id' must be integer, got %T", v)
}

File: sdks/php/src/Transport/Codec.php
<?php

declare(strict_types=1);

namespace Nexus\SDK\Transport;

use MessagePack\MessagePack;

/**
 * Wire codec for the Nexus RPC protocol.
 *
 * The server-side types live in nexus_protocol::rpc::types. rmp-serde
 * encodes Rust enums using the externally-tagged representation by
 * default:
 *
 *   - Unit variants (NexusValue::Null) → the string "Null".
 *   - Data-bearing variants (NexusValue::Str("hi")) → {"Str": "hi"}.
 *   - Result<T, E> → {"Ok": v} / {"Err": s}.
 *
 * toWire / fromWire translate between NexusValue and the on-wire
 * shape. encodeRequestFrame / decodeResponseBody handle the
 * `u32 LE length prefix + MessagePack body` frame format.
 */
final class Codec
{
    public static function toWire(NexusValue $v): mixed
    {
        return match ($v->kind) {
            NexusValueKind::Null => 'Null',
            NexusValueKind::Bool => ['Bool' => (bool) $v->value],
            NexusValueKind::Int => ['Int' => (int) $v->value],
            NexusValueKind::Float => ['Float' => (float) $v->value],
            NexusValueKind::Bytes => ['Bytes' => (string) $v->value],
            NexusValueKind::Str => ['Str' => (string) $v->value],
            NexusValueKind::Array => ['Array' => array_map(
                fn (NexusValue $e) => self::toWire($e),
                $v->value,
            )],
            NexusValueKind::Map => ['Map' => array_map(
                fn (array $pair) => [self::toWire($pair[0]), self::toWire($pair[1])],
                $v->value,
            )],
        };
    }

    public static function fromWire(mixed $raw): NexusValue
    {
        if ($raw === null) {
            return NexusValue::null();
        }
        if ($raw === 'Null') {
            return NexusValue::null();
        }
        if (is_bool($raw)) {
            return NexusValue::bool($raw);
        }
        if (is_int($raw)) {
            return NexusValue::int($raw);
        }
        if (is_float($raw)) {
            return NexusValue::float($raw);
        }
        if (is_string($raw)) {
            return NexusValue::str($raw);
        }
        if (!is_array($raw)) {
            throw new \RuntimeException(
                sprintf('decode: unexpected NexusValue wire type %s', get_debug_type($raw)),
            );
        }

        // Tagged map (associative array with a single key).
        if (self::isAssoc($raw)) {
            if (count($raw) !== 1) {
                throw new \RuntimeException(sprintf(
                    'decode: expected single-key tagged NexusValue, got %d keys',
                    count($raw),
                ));
            }
            $tag = array_key_first($raw);
            /** @var mixed $payload */
            $payload = $raw[$tag];
            return self::fromTagged((string) $tag, $payload);
        }

        // Plain list — treat as Array variant.
        $out = [];
        foreach ($raw as $e) {
            $out[] = self::fromWire($e);
        }
        return NexusValue::array($out);
    }

    /**
     * @param array<mixed> $arr
     */
    private static function isAssoc(array $arr): bool
    {
        return array_keys($arr) !== range(0, count($arr) - 1);
    }

    private static function fromTagged(string $tag, mixed $payload): NexusValue
    {
        switch ($tag) {
            case 'Null':
                return NexusValue::null();
            case 'Bool':
                return NexusValue::bool((bool) $payload);
            case 'Int':
                if (!is_int($payload) && !is_float($payload)) {
                    throw new \RuntimeException('decode: Int payload must be numeric');
                }
                return NexusValue::int((int) $payload);
            case 'Float':
                if (!is_int($payload) && !is_float($payload)) {
                    throw new \RuntimeException('decode: Float payload must be numeric');
                }
                return NexusValue::float((float) $payload);
            case 'Bytes':
                if (!is_string($payload)) {
                    throw new \RuntimeException('decode: Bytes payload must be a string');
                }
                return NexusValue::bytes($payload);
            case 'Str':
                return NexusValue::str((string) $payload);
            case 'Array':
                if (!is_array($payload)) {
                    throw new \RuntimeException('decode: Array payload must be array');
                }
                $out = [];
                foreach ($payload as $e) {
                    $out[] = self::fromWire($e);
                }
                return NexusValue::array($out);
            case 'Map':
                if (!is_array($payload)) {
                    throw new \RuntimeException('decode: Map payload must be array');
                }
                $pairs = [];
                foreach ($payload as $pair) {
                    if (!is_array($pair) || count($pair) !== 2) {
                        throw new \RuntimeException('decode: Map entry must be [key, value] pair');
                    }
                    $values = array_values($pair);
                    $pairs[] = [self::fromWire($values[0]), self::fromWire($values[1])];
                }
                return NexusValue::map($pairs);
        }
        throw new \RuntimeException(sprintf("decode: unknown NexusValue tag '%s'", $tag));
    }

    /**
     * Encode a request into a length-prefixed MessagePack frame.
     * Layout: u32_le(body_len) ++ msgpack(body).
     *
     * @param NexusValue[] $args
     */
    public static function encodeRequestFrame(int $id, string $command, array $args): string
    {
        $body = MessagePack::pack([
            'id' => $id,
            'command' => $command,
            'args' => array_map(fn (NexusValue $a) => self::toWire($a), $args),
        ]);
        return pack('V', strlen($body)) . $body;
    }

    /**
     * Decode a response body (MessagePack bytes AFTER the length prefix).
     *
     * @return array{id: int, ok: bool, value: ?NexusValue, err: string}
     */
    public static function decodeResponseBody(string $body): array
    {
        /** @var mixed $raw */
        $raw = MessagePack::unpack($body);
        if (!is_array($raw)) {
            throw new \RuntimeException(
                sprintf('decode: response must be a map, got %s', get_debug_type($raw)),
            );
        }
        if (!array_key_exists('id', $raw)) {
            throw new \RuntimeException('decode: response missing id');
        }
        $id = (int) $raw['id'];
        if (!array_key_exists('result', $raw) || !is_array($raw['result'])) {
            throw new \RuntimeException('decode: response missing result');
        }
        /** @var array<string, mixed> $result */
        $result = $raw['result'];
        if (count($result) !== 1) {
            throw new \RuntimeException('decode: Result must be a single-key tagged map');
        }
        $tag = (string) array_key_first($result);
        $payload = $result[$tag];
        return match ($tag) {
            'Ok' => ['id' => $id, 'ok' => true, 'value' => self::fromWire($payload), 'err' => ''],
            'Err' => ['id' => $id, 'ok' => false, 'value' => null, 'err' => (string) $payload],
            default => throw new \RuntimeException(sprintf("decode: Result must be 'Ok' or 'Err', got '%s'", $tag)),
        };
    }
}
